src/analyses/ai.h

```diff
@@ -28,10 +61,60 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "ai_storage.h"
 #include "is_threaded.h"
 
-/// The basic interface of an abstract interpreter. This should be enough
-/// to create, run and query an abstract interpreter.
+/// This is the basic interface of the abstract interpreter with default
+/// implementations of the core functionality.
+///
+/// Users of abstract interpreters should use the interface given by this class.
+/// It breaks into three categories:
+///
+/// 1. Running an analysis, via
+///    \ref ai_baset#operator()(const irep_idt&,const goto_programt&, <!--
+///    --> const namespacet&),
+///    \ref ai_baset#operator()(const goto_functionst&,const namespacet&)
+///    and \ref ai_baset#operator()(const abstract_goto_modelt&)
+/// 2. Accessing the results of an analysis, by looking up the history objects
+///    for a given location \p l using
+///    \ref ai_baset#abstract_traces_before(locationt)const
+///    or the domains using
+///    \ref ai_baset#abstract_state_before(locationt)const
+/// 3. Outputting the results of the analysis; see
+///    \ref ai_baset#output(const namespacet&, const irep_idt&,
+///    const goto_programt&, std::ostream&)const et cetera.
+///
+/// Where possible, uses should be agnostic of the particular configuration of
+/// the abstract interpreter.  The "tasks" that goto-analyze uses are good
+/// examples of how to do this.
+///
+/// From a development point of view, there are several directions in which
```

Comment:
I wonder if parts of this documentation wouldn't be better suited for a README file.

Reply:
Maybe.  At some point I want to write a paper describing how this works in detail, the interfaces, etc.  The goal of this commit was to keep the documentation up to date and keep the "if you want to do X then you should ..." comments.  Before that we need to have it merged and working!